File: 프로그래머스/lv2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;
class Solution {
    boolean solution(String str) {
        boolean answer = false;
        // 스택에 아무것도 없는데 )이게 들어오거나 마지막에 스택에 뭔가 남아있으면 false
        // (이게 들어오면 스택에 넣기. )이게 들어올떄 스택에 값이 있으면 pop
        Stack<Character> stack = new Stack<>();
        
        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            if (c == '(') {
                stack.push(c);
            }
            else {
                if (!stack.isEmpty()) {
                    stack.pop();
                }
                else {
                    return answer;
                }
            }
        }
        if (!stack.isEmpty()) {
            return answer;
        }
        answer = true;
        return answer;
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    answer = True
    
    # )가 올때 앞의 값이 있으면 뽑기
    # 비어있는데 )가 오면 False
    # 맨 마지막에 안비어있으면 False
    
    stack = []
    
    for cur in s:
        if cur == ')':
            if len(stack) == 0:
                answer = False
                break
            stack.pop()
        
        else:
            stack.append('(')
    
    if len(stack) > 0:
        answer = False
        
    return answer